src/api/submitOrder.js

```diff
@@ -1,45 +1,33 @@
 const { post } = require('request-promise')
-const validAssertions = require('../lib/validators/validateAssertions')
+const FP = require('lodash/fp')
+// TODO: define a schema for data and validate it.
+module.exports = async (dvf, orderData) => {
+  // allow passing in ethAddress (useful for testing).
+  const ethAddress = orderData.ethAddress || dvf.get('account')
 
-module.exports = async (
-  dvf,
-  gid,
-  cid,
-  partnerId,
-  feeRate = 0.0025,
-  dynamicFeeRate,
-  orderMetaData
-) => {
-  const ethAddress = dvf.get('account')
-  const type = 'EXCHANGE LIMIT'
-  const protocol = 'stark'
-  const data = {
-    cid,
-    gid,
-    type,
-    symbol: orderMetaData.symbol,
-    amount: orderMetaData.amount,
-    price: orderMetaData.price,
-    meta: {},
-    protocol,
-    partnerId,
-    feeRate,
-    dynamicFeeRate
+  data = {
+    type: 'EXCHANGE LIMIT',
+    protocol: 'stark',
+    feeRate: 0.0025,
+    ...FP.pick(
+      [
+        'amount',
+        'cid',
+        'dynamicFeeRate',
+        'feeRate',
+        'gid',
+        'partnerId',
+        'price',
+        'symbol'
+      ],
+      orderData
+    ),
+    meta: {
+      ethAddress,
+      ...(await dvf.createOrderMetaData(orderData))
+    }
   }
-
-  data.meta = {
-    ethAddress,
-    starkPublicKey: orderMetaData.starkPublicKey,
-    starkOrder: orderMetaData.starkOrder,
-    starkMessage: orderMetaData.starkMessage,
-    starkSignature: orderMetaData.starkSignature
-  }
-
   const url = dvf.config.api + '/v1/trading/w/submitOrder'
 
-  const submitResponse = await post(url, { json: data })
-
-  await dvf.getUserConfig()
-
-  return submitResponse
+  return post(url, { json: data })
```

Comment:
looks like this is equivalent to https://github.com/DeversiFi/dvf-client-js/blob/dev/src/api/submitOrderNew.js (but with code from https://github.com/DeversiFi/dvf-client-js/blob/dev/src/lib/dvf/createOrderPayload.js inlined). Could we not just use the same implementation as `submitOrderNew` (calling `createOrderPayload`)?

Reply:
CreateOrderMetaData will call ledger to sign order message if ledgerPath is present as we discussed. I have merged submitOrder, CreateOrderMetaData, and createOrderPayload into submitOrder. If we want to keep createOrderPayload then submitOrder will not do much except call createOrderPayload and call pub api. in the current implementation, there are 2 fewer methods and less replication. we can do a further iteration if required when we implement this for deposit and withdrawal and move the stark pvt key to dvf config.

Reply:
The main reason for having createOrderPayload separate was because it was used by the front end for generating the order, prior to signing it. This still may be useful, for example when adding wallets which are not keystore or Ledger

@Matko95 can probably contribute on whether this is necessary

Reply:
and regardless of the front ends constrains, I need createOrderPayload to create payloads for stark-trading api tests ([see here](https://github.com/ethfinex/launch-deversifi/blob/48d7a884789112aaabb50a040aa17f6c7bbb1db7/dvf-stark-trading/src/routes/api/v1/submitOrder.test.js#L63)). But it only makes sense to use it in tests if its actually used on the client when submitting an order.

Reply:
Got it 👍 createOrderPayload has now been separated out and submitOrder now calls createOrderPayload.